storage/memorybysubnet/peer_store.go

```diff
@@ -0,0 +1,768 @@
+
+// populateProm aggregates metrics over all shards and then posts them to
+// prometheus.
+func (ps *peerStore) populateProm() {
+	var numInfohashes, numSeeders, numLeechers uint64
+
+	for _, s := range ps.shards {
+		s.RLock()
+		numInfohashes += uint64(len(s.swarms))
+		numSeeders += s.numSeeders
+		numLeechers += s.numLeechers
+		s.RUnlock()
+	}
+
+	promInfohashesCount.Set(float64(numInfohashes))
+	promSeedersCount.Set(float64(numSeeders))
+	promLeechersCount.Set(float64(numLeechers))
+}
+
+func (ps *peerStore) getClock() int64 {
+	return atomic.LoadInt64(&ps.clock)
+}
+
+func (ps *peerStore) setClock(to int64) {
+	atomic.StoreInt64(&ps.clock, to)
+}
+
+func (ps *peerStore) shardIndex(infoHash bittorrent.InfoHash, af bittorrent.AddressFamily) uint32 {
+	// There are twice the amount of shards specified by the user, the first
+	// half is dedicated to IPv4 swarms and the second half is dedicated to
+	// IPv6 swarms.
+	idx := binary.BigEndian.Uint32(infoHash[:4]) % (uint32(len(ps.shards)) / 2)
+	if af == bittorrent.IPv6 {
+		idx += uint32(len(ps.shards) / 2)
+	}
+	return idx
+}
+
+func (ps *peerStore) PutSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	shard := ps.shards[ps.shardIndex(ih, p.IP.AddressFamily)]
+	shard.Lock()
+
+	if _, ok := shard.swarms[ih]; !ok {
+		shard.swarms[ih] = swarm{
+			seeders:  make(map[peerSubnet]map[serializedPeer]int64),
+			leechers: make(map[peerSubnet]map[serializedPeer]int64),
+		}
+	}
+
+	preferredSubnet := newPeerSubnet(p.IP, ps.ipv4Mask, ps.ipv6Mask)
+
+	// Allocate a new map if necessary.
+	if shard.swarms[ih].seeders[preferredSubnet] == nil {
+		shard.swarms[ih].seeders[preferredSubnet] = make(map[serializedPeer]int64)
+	}
+
+	// If this peer isn't already a seeder, update the stats for the swarm.
+	if _, ok := shard.swarms[ih].seeders[preferredSubnet][pk]; !ok {
+		shard.numSeeders++
+	}
+
+	// Update the peer in the swarm.
+	shard.swarms[ih].seeders[preferredSubnet][pk] = ps.getClock()
+
+	shard.Unlock()
+	return nil
+}
+
+func (ps *peerStore) DeleteSeeder(ih bittorrent.InfoHash, p bittorrent.Peer) error {
+	select {
+	case <-ps.closed:
+		panic("attempted to interact with stopped memory store")
+	default:
+	}
+
+	pk := newPeerKey(p)
+
+	shard := ps.shards[ps.shardIndex(ih, p.IP.AddressFamily)]
+	shard.Lock()
+
+	if _, ok := shard.swarms[ih]; !ok {
+		shard.Unlock()
+		return storage.ErrResourceDoesNotExist
+	}
+
+	preferredSubnet := newPeerSubnet(p.IP, ps.ipv4Mask, ps.ipv6Mask)
+	if _, ok := shard.swarms[ih].seeders[preferredSubnet][pk]; !ok {
+		shard.Unlock()
+		return storage.ErrResourceDoesNotExist
+	}
+
+	shard.numSeeders--
+	delete(shard.swarms[ih].seeders[preferredSubnet], pk)
+
+	if shard.swarms[ih].lenSeeders()|shard.swarms[ih].lenLeechers() == 0 {
+		delete(shard.swarms, ih)
+	} else if shard.swarms[ih].lenSeeders() == 0 {
```

Comment:
Oh has this been here all the time? Ignore my comment above then.
I think we should do `len(shard.swarms[ih].seeders[preferredSubnet] == 0` to clean up this subnet when it's empty. Otherwise we'd leave it as long as there is any subnet with at least one seeder left, doesn't feel right.